### ISSUE TITLE
Following instructions in CONTRIBUTING and README doesn't lead to expected results

### ISSUE THREAD

**Issue by gaurav7019** [openzim/mwoffliner#1590]
Installed all the dependencies, still getting this build error.

![Screenshot 2022-02-12 132827](https://user-images.githubusercontent.com/45741026/153702797-6c93ada4-5e02-46a7-99fe-db4a54719f62.png)

using Ubuntu 20.04 on WSL 2 (installed on Windows 11 Pro). Also, tried using installing Ubuntu 20.04 on a remote EC2 instance, getting the same error, so maybe it's not because of WSL.

Find the complete logs attached

[2022-02-12T07_57_19_753Z-debug.log](https://github.com/openzim/mwoffliner/files/8053030/2022-02-12T07_57_19_753Z-debug.log)

**Comment by kelson42** [comment]
I don't understand what you do. Can you share the full log (with the command you run) please?

**Comment by gaurav7019** [comment]
> Find the complete logs attached
> 
> [2022-02-12T07_57_19_753Z-debug.log](https://github.com/openzim/mwoffliner/files/8053030/2022-02-12T07_57_19_753Z-debug.log)

You can find the complete logs attached here. I ran the following command inside the project folder:

`npm ci`

after cloning the repo. I'm using the following versions of node and npm:

node v14.19.0
npm v6.14.16

**Comment by gaurav7019** [comment]
So, should I try with Virtual box only? I have an Ubuntu machine on my AWS ec2 instance as well, getting some error there as well, should I explore that further or go with the virtual box? Actually, I tried the virtual box thing before but it hangs a lot that way due to lack of graphic acceleration I guess.

**Comment by Jaifroid** [comment]
Ah yes, I remember! For us it was a severe dependency problem, and didn't work on WSL Ubuntu. It might be worth seeing if the dependency issues also occur on Hyper-V or Virtual Box. Ubuntu runs fine on Hyper-V (using a Generation 2 machine, 2 virtual processors, install it from a mounted ISO into 64GB VHDX). It may be better supported on Windows 11, though I don't know specifically about graphics performance -- it'll never be great virtualized. I'm sure Virtual Box is fine.

**Comment by Jaifroid** [comment]
> But, I think we can work on the bug for making it work with WSL as well.

That would be brilliant, if possible. I agree WSL Linux distros are becoming very popular because it gives near-native performance at least with its own FS (less so if accessing lots of files on one of the ntfs mountpoints like `/mnt/c/`, which can be slow with big npm install operations).

**Comment by kelson42** [comment]
@gaurav7019 It would be indeed great if we find a way to get MWoffliner working on WSL.

**Comment by gaurav7019** [comment]
I'll try exploring this alongside, let me know if you've any leads already that can help.

**Comment by kelson42** [comment]
@juuz0 said
   
I was able to install it correctly (except some deprecated warnings) on my WSL (Ubuntu 20.04, Windows 11). Commands I followed:

sudo apt update
sudo apt install redis-server
sudo apt install libjpeg-dev autoconf automake gcc
sudo apt install nodejs 
sudo apt install npm
git clone https://github.com/openzim/mwoffliner.git
cd mwoffliner
sudo npm i -g mwoffliner (this failed)
sudo npm cache clean --force (important)
sudo npm i -g mwoffliner --unsafe-perm
mwoffliner --help (yay!)

**Comment by Jaifroid** [comment]
@juuz0 I tried those steps, and got a long way, but it fell over again on `sudo npm i -g mwoffliner --unsafe-perm`. Screenshot showing commands from `git clone` onwards. This is Ubuntu 20.04.3 LTS on Windows 11 running in terminal.

![image](https://user-images.githubusercontent.com/4304337/153749837-dc5c2ada-4435-4fc9-8b4e-6dc16b3de2db.png)

**Comment by Jaifroid** [comment]
So I suppose ssh access is somehow not configured, even though git can clone the repo fine.

**Comment by kelson42** [comment]
We have a few dependence repos available via SSH and related to the umaintained Wikimedia packages.This is causing more and more problems looks like, see for example https://github.com/openzim/mwoffliner/issues/1592 open yesterday.

I have no real easy solution, except setup the client ssh infrastructure in WSL (shoukd be doable). Middle term, we should remove all this code (which is responsible for running a local Parsoid in case the upstream backend id old. This as well allow to have mobile output if not configured upstream).

**Comment by juuz0** [comment]
I had that error too. I followed this:
https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent

Related:
https://docs.github.com/en/authentication/troubleshooting-ssh/error-permission-denied-publickey

**Comment by Jaifroid** [comment]
I've managed to get past the ssh error by setting up keypair and adding to GitHub (warning to anyone following this, do not use a passphrase with the key because the npm script can't cope with that). But now a new error:

![image](https://user-images.githubusercontent.com/4304337/153751451-39147fa1-80a6-488f-8061-aacb578a2b8b.png)

**Comment by kelson42** [comment]
By investigating #1592, I had such an issue yesterday on GNU/Linux. I believe this is tied to "-g" part of the command. Maybe this ticket might help to understand better what is going on https://github.com/npm/cli/issues/624.

**Comment by Jaifroid** [comment]
OK, thanks. The only solution I could see there was not to run this as sudo. If I don't run as sudo, then that particular error goes away, but then I get later permission errors, e.g. for EACCES.

Only other option I could see in that thread was to move the clone of the repo out of user directory and start again. I'll try that later, though I have my doubts because this install isn't supposed to depend on the cloned Repo as @gaurav7019 mentioned. Presumably it updates the global npm config, though I don't know if that is also per-user.

**Comment by gaurav7019** [comment]
Using `--unsafe-perm` flag with `sudo npm i -g mwoffliner` works well for me on WSL (without cloning the repo as I am installing the package globally). But, for dev purposes, cloning the repo and building it inside it's folder using `nom ci --unsafe-perm` throws a build error attached below.

![image](https://user-images.githubusercontent.com/45741026/153761956-3f5f6e66-bfff-45b7-b95c-b266e3285961.png)

@Jaifroid Please confirm if you're able to reproduce this. So, we can move forward in this direction.

**Comment by gaurav7019** [comment]
> Using `--unsafe-perm` flag with `sudo npm i -g mwoffliner` works well for me on WSL

It didn't work earlier because I was using nvm for nodejs and npm. So, sudo was not working with npm and node. Fixed that to make it work fine.

**Comment by gaurav7019** [comment]
As I mentioned earlier, I spinned up a remote Ubuntu machine on AWS, SSHed into it from my vscode and tried building the repo there using `npm ci`. And, I got the same build error as mentioned above. So, I wonder, is it really related to WSL or are we missing something here.

**Comment by gaurav7019** [comment]
https://stackoverflow.com/a/47143457

As per this answer, this error could arise from resource limitations. Both my AWS instance (free tier, so fewer resources) and WSL might be having this. So, maybe this is the right direction. I am going to explore this for now.

**Comment by Jaifroid** [comment]
I need to check, I'm currently on a different machine. Will get back to you.

**Comment by Jaifroid** [comment]
So you never saw the permission errors above about not being able to create directories? I don't think it's a resources error, it's clearly a permissions error., probably a conflict between git expecting a user and root not being a normal user. Could you tell me which versions of node and npm worked for you? I am also trying to install globally.

**Comment by juuz0** [comment]
I was able to build it locally just fine by:
```
sudo uninstall -g mwoffliner
sudo cache clean --force
git clone https://github.com/openzim/mwoffliner.git
cd mwoffliner
npm ci
```

![image](https://user-images.githubusercontent.com/26814569/153770026-28784145-4882-45f0-9cef-e1d8b7c4f48a.png)

That said, I can't find the `mwoffliner` binary on installing locally. Expected it to be in `node_modules/.bin` but it wasn't there. Not sure if it's installation related.

**Comment by gaurav7019** [comment]
yes @Jaifroid, yours seems to be an error due to an older version of npm and node. The node version in Ubuntu 20.04 repo is v10 (which is very old). You should try installing node v14.19.0 (which comes with npm v6.14.16).

You can use nvm to manage your node versions. Firstly you should remove the current node by `sudo apt remove nodejs`. 
Then follow the following steps.

 - Install nvm using `curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.1/install.sh | bash` (Close the terminal and open again).
 - `nvm install 14.19.0`

Still, you need to create symbolic links from the versions of node and npm to `/usr/local/bin`. This is to make sudo work with node and npm as the node is not directly installed.

To create the req symbolic links:

`sudo ln -s "$NVM_DIR/versions/node/$(nvm version)/bin/node" "/usr/local/bin/node"`
`sudo ln -s "$NVM_DIR/versions/node/$(nvm version)/bin/npm" "/usr/local/bin/npm"`

After this you can follow the following steps to install mwoffliner globally:

- sudo npm cache clean --force
- sudo npm i -g mwoffliner --unsafe-perm

P.S: v14 is not the latest version of node but you should use that as mwoffliner is using lockfileVersion as 1 in the package-lock.json. And, npm 6 is the last npm version that works with that which comes with node v14.19.0. Otherwise, the installation will go under a one-time fix of changing the lockfileVersion in package-lock.json which might introduce unnecessary changes in the diff when you push to remote. @kelson42 please confirm this.

**Comment by juuz0** [comment]
@gaurav7019 I have got it working perfectly fine locally, can you please try these (exact order):
```
sudo rm -r mwoffliner
sudo uninstall -g mwoffliner
sudo cache clean --force
git clone https://github.com/openzim/mwoffliner.git
cd mwoffliner
npm ci
```

**Comment by gaurav7019** [comment]
> @gaurav7019 I have got it working perfectly fine locally, can you please try these (exact order):
> ```
> sudo rm -r mwoffliner
> sudo uninstall -g mwoffliner
> sudo cache clean --force
> git clone https://github.com/openzim/mwoffliner.git
> cd mwoffliner
> npm ci
> ```

Yes, on it already.

**Comment by kelson42** [comment]
I still wonder why you don't have the ssh errors if run from git on WSL.

**Comment by gaurav7019** [comment]
Are you talking about when we do `git clone`?

**Comment by Jaifroid** [comment]
Thank you @gaurav7019 - very clear instructions. I'll follow them when I next get a chance, so we can confirm the process for WSL users.

@kelson42 The ssh issue is solved by adding a keypair and uploading the public key to GitHub as @juuz0 pointed out. This will need to be in the instructions too.

**Comment by gaurav7019** [comment]
> sudo rm -r mwoffliner

@juuz0 What's this for? Is it for deleting the git repo cloned earlier?

Also, on doing,

> sudo uninstall -g mwoffliner

I'm getting `sudo: uninstall: command not found`. Any leads here?

> sudo cache clean --force

Here, you mean `sudo npm cache clean --force`, right?

**Comment by gaurav7019** [comment]
> @kelson42 The ssh issue is solved by adding a keypair and uploading the public key to GitHub as @juuz0 pointed out. This will need to be in the instructions too.

Yes, I confirm this as I had this configured already, so I didn't face this issue.

**Comment by juuz0** [comment]
> @juuz0 What's this for? Is it for deleting the git repo cloned earlier?

Yes

> I'm getting sudo: uninstall: command not found. Any leads here?

Sorry, I missed npm. `sudo npm uninstall -g mwoffliner`

> Here, you mean sudo npm cache clean --force, right?

Yes, forgot again

**Comment by gaurav7019** [comment]
@juuz0 Still getting the same build error on installing locally via git repo. (followed steps given by you).

![image](https://user-images.githubusercontent.com/45741026/153772174-e506585f-ff8c-4065-bb71-338cc71f4d16.png)

Attaching the complete logs below:
[2022-02-13T19_50_15_504Z-debug.log](https://github.com/openzim/mwoffliner/files/8056511/2022-02-13T19_50_15_504Z-debug.log)

Did you set any special permissions for npm before in the config? Or, configured your WSL in some different way?

**Comment by c01dc0ffee** [comment]
I haven't tried building it locally, but I was able to install it by following [this guide](https://stackoverflow.com/a/61976982) and then downgrading from npm 17.6.0 to  npm 14.15.0 as per [this advice on airtable.com](https://community.airtable.com/t/unable-to-install-blocks-cli/31490/15) using `nvm install v14.15.0`. After that, `npm i -g mwoffliner` installed without any problem.

**Comment by ImportTaste** [comment]
> I haven't tried building it locally, but I was able to install it by following [this guide](https://stackoverflow.com/a/61976982) and then downgrading from npm 17.6.0 to npm 14.15.0 as per [this advice on airtable.com](https://community.airtable.com/t/unable-to-install-blocks-cli/31490/15) using `nvm install v14.15.0`. After that, `npm i -g mwoffliner` installed without any problem.

It seems you don't actually need to downgrade when using nvm. Because nvm doesn't use any root-controlled directories, it makes it so the latest version works as well.

**Comment by kelson42** [comment]
@c01dc0ffee This is good new. Would you be able to make a PR with and updated `README.md` explaining the excat whole procedure? So we could test it on many computer and see if we can advert it?

**Comment by Jaifroid** [comment]
@ImportTaste I followed your instructions, but I still get a major error with the global install:

![image](https://user-images.githubusercontent.com/4304337/158050089-f2f0f3eb-a782-405e-bacd-71edeaf2fa07.png)

**Comment by kelson42** [comment]
@Jaifroid Here it seems to ne we have a pure permission problem, your user is `jaifrod` and you try to write in the `root` HOME. I would recommend first to write the instruction to run mwoffliner locally and then have a look to the general install. To develop, only the first is needed.

**Comment by kelson42** [comment]
@Jaifroid Have you had the opportunity to retry again? It should really work meanwhile.

**Comment by benoit74** [comment]
Looks like nobody cares + probably fixed => closing until someone confirms there is still an issue we can fix
